File: P52/sol.py
def sum_digits(n):
   r = 0
   while n:
       r, n = r + n % 10, n // 10
   return r

for num in range(123456, 166666):
    if sum_digits(num) == sum_digits(2*num) and \
        sum_digits(num) == sum_digits(3*num) and \
        sum_digits(num) == sum_digits(3*num) and \
        sum_digits(num) == sum_digits(4*num) and \
        sum_digits(num) == sum_digits(5*num) and \
        sum_digits(num) == sum_digits(6*num):
        if sorted(str(num)) == sorted(str(2*num)) and \
            sorted(str(num)) == sorted(str(2*num)) and \
            sorted(str(num)) == sorted(str(3*num)) and \
            sorted(str(num)) == sorted(str(4*num)) and \
            sorted(str(num)) == sorted(str(5*num)) and \
            sorted(str(num)) == sorted(str(6*num)):
            print(num, 2*num, 3*num, 4*num, 5*num, 6*num)

File: P-sqrt13/sol.hs
module Main where

main = do
  print . sumDigits $ squareRoot (13 * 10 ^ 2000 :: Integer) - 3

sumDigits :: Integer -> Integer
sumDigits x
  | x < 10 = x
  | otherwise = x `mod` 10 + sumDigits (x `div` 10)

(^!) :: (Num a) => a -> Int -> a
(^!) x n = x ^ n

squareRoot :: Integer -> Integer
squareRoot 0 = 0
squareRoot 1 = 1
squareRoot n =
  let twopows = iterate (^! 2) 2
      (lowerRoot, lowerN) =
        last $ takeWhile ((n >=) . snd) $ zip (1 : twopows) twopows
      newtonStep x = div (x + div n x) 2
      iters = iterate newtonStep (squareRoot (div n lowerN) * lowerRoot)
      isRoot r = r ^! 2 <= n && n < (r + 1) ^! 2
   in head $ dropWhile (not . isRoot) iters